logging/google/cloud/logging/handlers/GAEHandler.py

```diff
@@ -0,0 +1,93 @@
+# limitations under the License.
+
+"""Logging handler for App Engine Flexible
+
+Logs to the well-known file that the fluentd sidecar container on App Engine
+Flexible is configured to read from and send to Stackdriver Logging.
+
+This file is largely copied from:
+
+https://github.com/GoogleCloudPlatform/python-compat-runtime/blob/master/appengine-vmruntime/vmruntime/cloud_logging.py
+
+"""
+
+import json
+import logging.handlers
+import math
+import os
+
+LOG_PATH_TEMPLATE = '/var/log/app_engine/app.{pid}.json'
+MAX_LOG_BYTES = 128 * 1024 * 1024
+LOG_FILE_COUNT = 3
+
+
+class GAEHandler(logging.handlers.RotatingFileHandler):
+    """A handler that writes to the GAE fluentd Stackdriver logging file.
+
+    Writes to the file that the fluentd agent on App Engine Flexbile is
+    configured to discover logs and send them to  Stackdriver Logging.
+    Log entries are wrapped in JSON and with appropriate metadata. The
+    process of converting the user's formatted logs into a JSON payload for
+    Stackdriver Logging consumption is implemented as part of the handler
+    itself, and not as a formatting step, so as not to interfere with
+    user-defined logging formats.
+
+    The handler will look for HTTP header information in the environment (which
+    will be present in the GAE 1.0-compatible runtime) and, if it exists, will
+    parse the X-Cloud-Trace-Context header to add a Trace ID to each log
+    record.
+    Logging calls can also alternatively 'trace_id' in as a field in the
+    'extra' dict, which will be used preferentially to fill out the Trace ID
+    metadata.
+    """
+
+    def __init__(self):
+        # Large log entries will get mangled if multiple workers write to the
+        # same file simultaneously, so we'll use the worker's PID to pick a log
+        # filename.
+        filename = LOG_PATH_TEMPLATE.format(pid=os.getpid())
+        super(GAEHandler, self).__init__(filename, maxBytes=MAX_LOG_BYTES,
+                                         backupCount=LOG_FILE_COUNT)
+
+    def format(self, record):
+        """Format the specified record default behavior, plus JSON and
+        metadata.
+
+        :type record: `logging.LogRecord`
+        :param record: the log record
+
+        :rtype: str
+        :returns: JSON str to be written to the log file
+        """
+        message = super(GAEHandler, self).format(record)
+
+        subsecond, second = math.modf(
+            record.created)
+
+        payload = {'message': message, 'timestamp': {'seconds': int(second),
+                                                     'nanos': int(
+                                                         subsecond * 1e9)},
+                   'thread': record.thread, 'severity': record.levelname}
+
+        trace_id = getattr(record, 'trace_id', None)
+
+        if not trace_id:
+            trace_id = os.getenv(
+                'HTTP_X_CLOUD_TRACE_CONTEXT', '').split('/')[0]
+
+        if trace_id:
```

Comment:
With the suggested change about, you could/should check `if trace_id is not None` here.